sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala

```diff
@@ -54,3 +60,367 @@ case class CurrentTimestamp() extends LeafExpression {
+
+  override protected def nullSafeEval(input: Any): Any = {
+    val (year, dayInYear) = calculateYearAndDayInYear(input.asInstanceOf[Int])
+    val leap = if (isLeapYear(year)) 1 else 0
+    dayInYear match {
+      case i: Int if i <= 31 => i
+      case i: Int if i <= 59 + leap => i - 31
+      case i: Int if i <= 90 + leap => i - 59 - leap
+      case i: Int if i <= 120 + leap => i - 90 - leap
+      case i: Int if i <= 151 + leap => i - 120 - leap
+      case i: Int if i <= 181 + leap => i - 151 - leap
+      case i: Int if i <= 212 + leap => i - 181 - leap
+      case i: Int if i <= 243 + leap => i - 212 - leap
+      case i: Int if i <= 273 + leap => i - 243 - leap
+      case i: Int if i <= 304 + leap => i - 273 - leap
+      case i: Int if i <= 334 + leap => i - 304 - leap
+      case i: Int => i - 334 - leap
+    }
+  }
+
+  override protected def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    nullSafeCodeGen(ctx, ev, days => {
+      codeGen(ctx, ev, days, (year, dayInYear) => {
+        val leap = ctx.freshName("leap")
+        s"""
+            int $leap = ($year % 4) == 0 && (($year % 100) != 0 || ($year % 400) == 0) ? 1 : 0;
+            if ($dayInYear <= 31) {
+              ${ev.primitive} = $dayInYear;
+            } else if ($dayInYear <= 59 + $leap) {
+              ${ev.primitive} = $dayInYear - 31;
+            } else if ($dayInYear <= 90 + $leap) {
+              ${ev.primitive} = $dayInYear - 59 - $leap;
+            } else if ($dayInYear <= 120 + $leap) {
+              ${ev.primitive} = $dayInYear - 90 - $leap;
+            } else if ($dayInYear <= 151 + $leap) {
+              ${ev.primitive} = $dayInYear - 120 - $leap;
+            } else if ($dayInYear <= 181 + $leap) {
+              ${ev.primitive} = $dayInYear - 151 - $leap;
+            } else if ($dayInYear <= 212 + $leap) {
+              ${ev.primitive} = $dayInYear - 181 - $leap;
+            } else if ($dayInYear <= 243 + $leap) {
+              ${ev.primitive} = $dayInYear - 212 - $leap;
+            } else if ($dayInYear <= 273 + $leap) {
+              ${ev.primitive} = $dayInYear - 243 - $leap;
+            } else if ($dayInYear <= 304 + $leap) {
+              ${ev.primitive} = $dayInYear - 273 - $leap;
+            } else if ($dayInYear <= 334 + $leap) {
+              ${ev.primitive} = $dayInYear - 304 - $leap;
+            } else {
+              ${ev.primitive} = $dayInYear - 334 - $leap;
+            }
+         """
+      })
+    })
+  }
+
+}
+
+case class WeekOfYear(child: Expression) extends UnaryExpression with ImplicitCastInputTypes {
+
+  override def dataType: DataType = IntegerType
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(DateType)
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String =
+    nullSafeCodeGen(ctx, ev, (time) => {
+      val cal = classOf[Calendar].getName
+      val c = ctx.freshName("cal")
+      s"""
+        $cal $c = $cal.getInstance();
```

Comment:
Reynold is it possible to use `addMutableState` here? After the discussions regarding thread safety I guess it's not possible to use. Am I right?